### PR TITLE
Added resilience4j wrapper to internal operation calls.

### PR DESCRIPTION
**Pull request by amanis22**

### CODE REVIEW

File: src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package com.sparetimedevs.suspendmongo.resilience
```
Comment (joramnv) [inline src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt:17]: Missing the copyright comment. You can actually configure your IDE to auto generate them for you.

File: src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    private val bulkheadConfig = BulkheadConfig.custom()
+            .maxConcurrentCalls(maxConcurrentCalls)
+            .maxWaitTime(maxWaitMillis)
+            .build()
+
+    private val retry = Retry.of("mongo-retry", retryConfig)
+    private val bulkhead = Bulkhead.of("mongo-bulkhead", bulkheadConfig)
+
+    fun <T> executeSupplier(supplier: Supplier<T?>): T? = bulkhead.executeSupplier { retry.executeSupplier(supplier) }
```
Comment (joramnv) [inline src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt]: Does this need to be `Supplier<nullable type>` and have a nullable return type? It leads to more questionmarks.

File: src/main/kotlin/com/sparetimedevs/suspendmongo/crud/InternalOperation.kt [hunk 3]
```diff
@@ -18,20 +18,24 @@ package com.sparetimedevs.suspendmongo.crud
...
 import com.sparetimedevs.suspendmongo.result.Error
 import com.sparetimedevs.suspendmongo.result.Result
 import kotlinx.coroutines.reactive.awaitFirstOrNull
 import kotlinx.coroutines.reactive.consumeEach
 import org.bson.conversions.Bson
+import java.util.function.Supplier
 
 //TODO create better error messages.
 
+private val resilience: Resilience by lazy { Resilience() }
```
Comment (joramnv) [inline src/main/kotlin/com/sparetimedevs/suspendmongo/crud/InternalOperation.kt]: Don’t think this is a smart way to initialise a dependency. But maybe I lag knowledge regarding lazy.
In any case, I do not see how a user of the library could set the Resilience with different properties. Maybe this dependency could be part of the Database object. Which can be created by the user of the library. That way the user does have control over how resilience the database calls need to be.

**Review by joramnv — COMMENTED** [review]
Overall looking good! Nice and clean. Left 3 comments for you.

File: src/main/kotlin/com/sparetimedevs/suspendmongo/crud/InternalOperation.kt (hunk 3 above)
Comment (joramnv) [reply src/main/kotlin/com/sparetimedevs/suspendmongo/crud/InternalOperation.kt]: `class Database(connectionString: String, name: String, val resilience: Resilience = Resilience()) { ... }`

File: src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt (hunk 1 above)
Comment (amanis22) [reply src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt:17]: Updating.

File: src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt (hunk 2 above)
Comment (amanis22) [reply src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt]: Removed the nullable type from the supplier. Seems like it's happy. Removed ? from the callers.

File: src/main/kotlin/com/sparetimedevs/suspendmongo/crud/InternalOperation.kt (hunk 3 above)
Comment (amanis22) [reply src/main/kotlin/com/sparetimedevs/suspendmongo/crud/InternalOperation.kt]: Check out this for more info regarding delegated properties. [how-kotlins-delegated-properties-and-lazy-initialization-work](https://medium.com/til-kotlin/how-kotlins-delegated-properties-and-lazy-initialization-work-552cbad8be60)

I'll take a look at refactoring to allow override of the resilience properties as per your suggestion. I'll push another update later!

Comment (joramnv) [reply src/main/kotlin/com/sparetimedevs/suspendmongo/crud/InternalOperation.kt]: Thanks for the article, good read. I guess it is a viable option if we want to initialise Resilience with default values only (for now).

**Pull request merged**